### PR TITLE
feat: parallelize nodes update

### LINKED ISSUE #1096: Enhancement: parallelize nodes update
**What would you like to be added**:
Currently, we are updating nodes in a sequential way.
https://github.com/kubernetes-sigs/node-feature-discovery/blob/821e042dbfde7fe509a26f2215385d7302d9fb51/pkg/nfd-master/nfd-master.go#L295-L306
I think that we can optimize this step by parallelizing this process.

**Why is this needed**:
Optimization of updating nodes process, in terms of execution time.

### PR DESCRIPTION
**Pull request by AhmedGrati**
Resolves #1096.

### CODE REVIEW

**Review by marquiz — COMMENTED** [review]
Thanks @AhmedGrati for the PR.

I think we need to throttle the max parallelism (i.e. how many concurrent go routines are generated) somehow. Say we have a cluster with 2000 nodes we don't probably want the fire 2000 goroutines at the same time

/milestone v0.14.0

**Comment by PiotrProkop** [comment]
Can we encorporate workqueue here? Have a flag than configures the number of consumers(which will update specific node) and we will just `workqueue.Add` nodes when we want to update them. 

We can use https://pkg.go.dev/k8s.io/client-go/util/workqueue for example.

**Comment by marquiz** [comment]
> Can we encorporate workqueue here?

Yeah, for example something like this

**Comment by AhmedGrati** [comment]
Hey @PiotrProkop, I've just come up with this design to include workqueue, and to have a goroutines pool responsible for labeling nodes.
NFD drawio
But at the same time @marquiz, suggested that we can use [SizedWaitGroup.](https://github.com/remeh/sizedwaitgroup)
I wanted to gather your input here, on what should we do next.
1. Keep it simple as is now, and add the SizedWaitGroup.
2. Implement a more sophisticated solution and add the `workqueue` package.

**Comment by AhmedGrati** [comment]
ping @marquiz @PiotrProkop

**Comment by PiotrProkop** [comment]
I think decoupling labeling into separate `workqueue` is better approach for the future. It would move labeling responsibility into separate component and we would have to just `Add` labeling request from the main loop.

On the other hand I know that implementing it is non trivial and I think `SizedWaitGroup` will also work but we would just have to make sure that we move all labeling into single `SizedWaitGroup`  or single `ErrGroup`.

@marquiz WDYT?

**Comment by marquiz** [comment]
> I think decoupling labeling into separate `workqueue` is better approach for the future. It would move labeling responsibility into separate component and we would have to just `Add` labeling request from the main loop.
> 
> On the other hand I know that implementing it is non trivial and I think `SizedWaitGroup` will also work but we would just have to make sure that we move all labeling into single `SizedWaitGroup` or single `ErrGroup`.

I agree with @PiotrProkop here. Using workqueues would be more future proof and architecturally cleaner for the long run.

However, as @PiotrProkop said it's not trivial and I think lot more work (including tests) than the simple-and-stupid SizedWaitGroup. Using SizedWaitGroup should be safe and simple (__currently__) as there is only one place (and goroutine) from where we do labeling.

Thus, I'm fine with either approach. You can choose 😉

**Comment by AhmedGrati** [comment]
Thanks for your input @marquiz @PiotrProkop, I'll implement it using `workqueue`.

**Comment by AhmedGrati** [comment]
PTAL @marquiz @PiotrProkop

**Comment by PiotrProkop** [comment]
> PTAL @marquiz @PiotrProkop

Instead of spinning workers on demand up to the limmit, you can precreate desired amount of workers and make them wait for `node` to be processed. You can take a look at this example: https://github.com/kubernetes/client-go/blob/master/examples/workqueue/main.go

**Comment by AhmedGrati** [comment]
@PiotrProkop I updated it with changes inspired from the example the you provided.

File: pkg/nfd-master/nfd-master.go
Comment (PiotrProkop) [inline pkg/nfd-master/nfd-master.go:173]: why use `delaying` queue when you are not using `AddAfter`? It would be better to use standatd queue:
`workqueue.New()` or rate limited one and make it only retry key/node some specified amount of times as in example I have pasted earlier(to prevent infinite processing of one failing key).

File: pkg/nfd-master/nfd-master.go
Comment (marquiz) [inline pkg/nfd-master/nfd-master.go:276]: nit: looks a bit strange that this is a separate method 🧐 How about just doing this error checking inline in `processNodeLabelRequest`

File: pkg/nfd-master/nfd-master.go
Comment (marquiz) [reply pkg/nfd-master/nfd-master.go:173]: Good point @PiotrProkop 👍 We don't want to infinitely retry, e.g. nodes that are deleted from the cluster

Comment (AhmedGrati) [reply pkg/nfd-master/nfd-master.go:173]: Yes makes sense thanks, do you suggest a max number of retries? @PiotrProkop @marquiz

File: pkg/nfd-master/nfd-master.go
Comment (PiotrProkop) [inline pkg/nfd-master/nfd-master.go:62]: I think 200 is kinda big number for default concurrent labelers, maybe we can set it to `10` ?

File: cmd/nfd-master/main.go
Comment (PiotrProkop) [inline cmd/nfd-master/main.go:159]: This should be `flagset.Uint` instead of `Int` cause negative values are incorrect.

Comment (AhmedGrati) [reply cmd/nfd-master/main.go:159]: Yes 100% true, but if the user inputs for example `-5` it would result in having `18446744073709551611 ` for the `MaxNumberOfConcurrentLabelers`. I would suggest working with `math.Abs` to handle negative values. wdyt?

File: deployment/helm/node-feature-discovery/values.yaml
Comment (PiotrProkop) [inline deployment/helm/node-feature-discovery/values.yaml:43]: s/200/10

**Comment by AhmedGrati** [comment]
@marquiz @PiotrProkop PTAL when you have time.

File: pkg/nfd-master/nfd-master.go
Comment (PiotrProkop) [inline pkg/nfd-master/nfd-master.go:306]: you can inline err definition:
```suggestion
	if err := m.nfdAPIUpdateOneNode(nodeName.(string)); err != nil {
```

**Review by PiotrProkop — COMMENTED** [review]
small nit otherwise LGTM, will wait for proper approval one @marquiz will review this PR.

File: pkg/nfd-master/nfd-master.go
Comment (marquiz) [inline pkg/nfd-master/nfd-master.go:316]: I'd suggest to rename this to not being about labeling but about updating nodes `runNodeUpdater` or smth

File: pkg/nfd-master/nfd-master.go
Comment (marquiz) [inline pkg/nfd-master/nfd-master.go:321]: Also, I'd suggest to rename this, e.g. to `runNodeUpdaterPool`

File: pkg/nfd-master/nfd-master.go
Comment (marquiz) [reply pkg/nfd-master/nfd-master.go:325]: Actually, taking a closer look, why do we use `wait.Until()` in the first place as we don't utilize the periodic calling of the function, for example?

Wouldn't it be simpler to do just

```golang
		go runLabeler(stopChan)
```

And then handle the stop channedl in runLabeler

Comment (AhmedGrati) [reply pkg/nfd-master/nfd-master.go:325]: Makes sense, good point 👌

**Comment by AhmedGrati** [comment]
@marquiz Thank you for the review. It's updated now.

**Comment by AhmedGrati** [comment]
@marquiz any other comments on this?

File: pkg/nfd-master/nfd-master.go
Comment (marquiz) [inline pkg/nfd-master/nfd-master.go:330]: Now that I'm looking at this I think it doesn't work at all: processNodeLabelRequest blocks and in practice the stop signal is probably never received (in particular as m.nodeUpdaterPoolStop is re-initialized before we get to the select again) -> growing number of leaking goroutines on every re-configuration event -> bad

I think the right thing to do is to shutdown the queue in runNodeUpdaterPool so that all blocking get operations get cancelled.
```suggestion
	for m.processNodeLabelRequest() {
	}
```

File: pkg/nfd-master/nfd-master.go
Comment (marquiz) [inline pkg/nfd-master/nfd-master.go:301]: Safer to use local variable as we operate in different  goroutines and the queue is our "sync channel"

```suggestion
func (m *nfdMaster) processNodeLabelRequest(queue workqueue.RateLimitingInterface) bool {
```

File: pkg/nfd-master/nfd-master.go
Comment (marquiz) [inline pkg/nfd-master/nfd-master.go:348]: Hmm, I think ideally we should wait for the updater goroutines to finish (waitgroup). To be on the safe side, even if an unlikely race. We might e.g. have in-flight retries that get intermixed with node updates from the new pool after re-configuration.

One idea: `runNodeUpdaterPool()` could be changed to non-blocking `startNodeUpdaterPool` that just creates the worker goroutines and adds them to a waitgroup and returns. Then `stopNodeUpdaterPool` would signal these worker goroutines to stop and wait for them to finish (with the waitgroup). 

Also, I don't think we actually need the stop channel for anything. 

```suggestion
	m.queue.ShutDown()
```

**Review by marquiz — COMMENTED** [review]
Some more comments. I think there's still some work to do on the goroutine management front

**Review by marquiz — COMMENTED** [review]
Thanks for the update. A few more comments. @PiotrProkop could you also take a look.

Also @AhmedGrati you added some benchmarks. Have you run them? Do you have any data? Results could be added into the commit message as informational

**Comment by AhmedGrati** [comment]
Thanks for the review @marquiz. 
For the benchmarking, in the old implementation, it updates `1000` nodes (ofc we are mocking the communication with k8s API) in a duration that goes between `500 micro-seconds` and `650 micro-seconds`. In the current implementation, the duration is between `300 micro-seconds` and `450 micro-seconds`. So I don't know whether It's worth mentioning in the description of the PR or not.
However, I still believe that macro-benchmarking would produce more realistic results regarding this improvement. WDYT?

**Review by marquiz — COMMENTED** [review]
Nice work @AhmedGrati. I think we could merge this already.

Although now that I took another comprehensive look at this I think it would make sense to separate the node updater pool into a separate type. I think it would make it cleaner, architecturally, also clearly separating the data.

```golang
type nodeUpdaterPool struct {
	queue                workqueue.RateLimitingInterface
	wg                   sync.WaitGroup
        nfdMaster            nfdMaster
}

func (u *nodeUpdaterPool) runNodeUpdater()

func (u *nodeUpdaterPool) start(parallelism int)

func (u *nodeUpdaterPool) stop()
```

/assign @PiotrProkop

**Review by marquiz — COMMENTED** [review]
Nice work @AhmedGrati. I think we could merge this already.

Although now that I took another comprehensive look at this I think it would make sense to separate the node updater pool into a separate type. I think it would make it cleaner, architecturally, also clearly separating the data.

```golang
type nodeUpdaterPool struct {
	queue                workqueue.RateLimitingInterface
	wg                   sync.WaitGroup
        nfdMaster            nfdMaster
}

func (u *nodeUpdaterPool) runNodeUpdater()

func (u *nodeUpdaterPool) start(parallelism int)

func (u *nodeUpdaterPool) stop()
```

In `start()` you could also guard against calling it twice (without calling stop() in between)

/assign @PiotrProkop

**Comment by AhmedGrati** [comment]
Nice suggestion, I will work on that.
/hold

File: pkg/nfd-master/node-updater-pool.go
Comment (marquiz) [inline pkg/nfd-master/node-updater-pool.go:19]: This singleton approach looks a bit weird as you give nfdMaster as an argument but that argument is totally omitted in subsequent calls.

How about changing this to

```suggestion
func newNodeUpdaterPool(nfdMaster nfdMaster) *nodeUpdaterPool
```

and then calling this in NewNfdMaster and storing the instance in the nfdMaster struct?

Comment (PiotrProkop) [reply pkg/nfd-master/node-updater-pool.go:19]: I agree, It would be much more readable, since the nodeUpdaterPool is only used from within `nfdMaster`.

File: pkg/nfd-master/node-updater-pool.go
Comment (PiotrProkop) [inline pkg/nfd-master/node-updater-pool.go:41]: and if this is the only place where `nfdAPIUpdateOneNode` is called you can just remove this func from `nfdMaster` and make it `nodeUpdaterPoll` func, this way you don't need to pass `nfdMaster` to `nodeUpdaterPool` but prolly just Kubernetes client.

File: pkg/nfd-master/node-updater-pool.go
Comment (AhmedGrati) [reply pkg/nfd-master/node-updater-pool.go:19]: Wouldn't that result in a circular dependency?

Comment (AhmedGrati) [reply pkg/nfd-master/node-updater-pool.go:19]: `nfdMaster contains nodeUpdaterPool` and `nodeUpdaterPool contains nfdMaster`.

Comment (marquiz) [reply pkg/nfd-master/node-updater-pool.go:19]: Think about it like doubly-linked list

File: pkg/nfd-master/node-updater-pool.go
Comment (marquiz) [inline pkg/nfd-master/node-updater-pool.go:16]: One suggestion to potentially beautify a bit

```suggestion
type nodeUpdaterPool struct {
	sync.Mutex
	workqueue.RateLimitingInterface
	
	wg        sync.WaitGroup
	nfdMaster *nfdMaster
	started   bool
}
```

With this you could do simpler `u.Lock()` instead of `u.mutex.Lock()` and e.g. `pool.Add(nodeName)` instead of `pool.queue.NodeName()`. WDYT?

File: pkg/nfd-master/node-updater-pool.go
Comment (marquiz) [inline pkg/nfd-master/node-updater-pool.go:57]: We could do away with the started bool with smth like

```suggestion
	if u.queue != nil && !u.queue.ShuttingDown() {
```

File: pkg/nfd-master/node-updater-pool.go
Comment (marquiz) [inline pkg/nfd-master/node-updater-pool.go:62]: Use simple "fieldName" type keys for the log data. Also start with lower-case letter

```suggestion
	klog.InfoS("starting the NFD master node updater pool", "parallelism", parallelism)
```

File: pkg/nfd-master/node-updater-pool.go
Comment (marquiz) [inline pkg/nfd-master/node-updater-pool.go:78]: nit: we've tried to maintain a convention to start log messages with lower-case

```suggestion
		klog.InfoS("the NFD master node updater pool is not running.")
```

ditto other messages, too...

File: pkg/nfd-master/node-updater-pool.go
Comment (AhmedGrati) [reply pkg/nfd-master/node-updater-pool.go:16]:
```go
	u.queue.ShutDown()
	u.wg.Wait()
```
Would be transformed to:
```go
	u.ShutDown()
	u.wg.Wait()
```
Isn't it misleading to have `u.Shutdown()` and then `u.wg.Wait()`?

Comment (AhmedGrati) [reply pkg/nfd-master/node-updater-pool.go:16]: It gives the intention that we're shutting down the pool and then waiting for the waitgroup, where in reality we're shutting down only the queue.

Comment (marquiz) [reply pkg/nfd-master/node-updater-pool.go:16]: Yeah, maybe it's cleaner to have queue as a separate field

File: pkg/nfd-master/node-updater-pool.go
Comment (marquiz) [inline pkg/nfd-master/node-updater-pool.go:71]:
```suggestion
```

File: pkg/nfd-master/node-updater-pool.go
Comment (marquiz) [inline pkg/nfd-master/node-updater-pool.go:75]:
```suggestion
	if u.queue == nil || u.queue.ShuttingDown() {
```

**Comment by AhmedGrati** [comment]
@marquiz any idea why is this failing?

File: pkg/nfd-master/nfd-master.go
Comment (marquiz) [inline pkg/nfd-master/nfd-master.go:1183]: I think we could simplify and remove the (wrongly named) const alltogether by just adding the default (10) to `newDefaultConfig()` and then banning zero.

```suggestion
	}
	if c.NfdApiParallelism <= 0 {
		return fmt.Errorf("the maximum number of concurrent labelers should be a non-zero positive number")
	}
```

WDYT?

Comment (AhmedGrati) [reply pkg/nfd-master/nfd-master.go]: Yes that's cleaner.

File: pkg/nfd-master/nfd-master.go
Comment (marquiz) [inline pkg/nfd-master/nfd-master.go:62]:
```suggestion
```

**Comment by marquiz** [comment]
@AhmedGrati also one with: could you clean up the commit message 😇

File: pkg/nfd-master/node-updater-pool.go
Comment (marquiz) [inline pkg/nfd-master/node-updater-pool.go:1]: Just realized that this file is missing the license/copyright header. Could you add that one?

Comment (AhmedGrati) [reply pkg/nfd-master/node-updater-pool.go:1]: Done

File: pkg/nfd-master/node-updater-pool.go
Comment (marquiz) [inline pkg/nfd-master/node-updater-pool.go:2]: nit 🙈 put 2023 here
```suggestion
Copyright 2023 The Kubernetes Authors.
```

**Review by marquiz — APPROVED** [review]
Good job and thanks for the persistence on this @AhmedGrati. I don't think I have anything to add for now -> approved

@PiotrProkop could you take look, with fresh eyes?

**Review by marquiz — APPROVED** [review]
Good job and thanks for the persistence on this @AhmedGrati. I don't think I have anything to add for now -> approved

@PiotrProkop could you take look, with fresh eyes?

File: pkg/nfd-master/node-updater-pool.go
Comment (PiotrProkop) [reply pkg/nfd-master/node-updater-pool.go:49]: Ok, we should create issue to track this work.

**Review by PiotrProkop — COMMENTED** [review]
/lgtm

**Pull request merged**
